govcd/api_token.go

```diff
@@ -21,122 +21,208 @@ import (
 }
 
-// GetBearerTokenFromApiToken uses an API token to retrieve a bearer token
-// using the refresh token operation.
-func (vcdClient *VCDClient) GetBearerTokenFromApiToken(org, token string) (*types.ApiTokenRefresh, error) {
-	if vcdClient.Client.APIVCDMaxVersionIs("< 36.1") {
-		version, err := vcdClient.Client.GetVcdFullVersion()
+// GetTokenByNameForCurrentUser retrieves a Token by name and username
+func (vcdClient *VCDClient) GetTokenByNameAndUsername(tokenName, userName string) (*Token, error) {
+	client := vcdClient.Client
+	if client.APIVCDMaxVersionIs("< 36.1") {
+		version, err := client.GetVcdFullVersion()
 		if err == nil {
 			return nil, fmt.Errorf("minimum version for API token is 10.3.1 - Version detected: %s", version.Version)
 		}
 		// If we can't get the VCD version, we return API version info
-		return nil, fmt.Errorf("minimum API version for API token is 36.1 - Version detected: %s", vcdClient.Client.APIVersion)
+		return nil, fmt.Errorf("minimum API version for API token is 36.1 - Version detected: %s", client.APIVersion)
+	}
+
+	queryParameters := url.Values{}
+	queryParameters.Add("filter", fmt.Sprintf("(name==%s;owner.name==%s;(type==PROXY,type==REFRESH))", tokenName, userName))
+
+	tokens, err := vcdClient.GetAllTokens(queryParameters)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get token by name and owner: %s", err)
 	}
-	var userDef string
-	newUrl := new(url.URL)
-	newUrl.Scheme = vcdClient.Client.VCDHREF.Scheme
-	newUrl.Host = vcdClient.Client.VCDHREF.Host
-	urlStr := newUrl.String()
+
+	token, err := oneOrError("name", tokenName, tokens)
+	if err != nil {
+		return nil, err
+	}
+
+	return token, nil
+}
+
+func (vcdClient *VCDClient) RegisterToken(org string, tokenParams *types.ApiTokenParams) (*types.ApiTokenParams, error) {
+	client := vcdClient.Client
+
+	// If the client is a user of an org, the endpoint is oauth/tenant/orgName/register
+	// if the client is a user of sysorg, the endpoint is oauth/provider/register
+	userDef := "tenant/" + org
 	if strings.EqualFold(org, "system") {
 		userDef = "provider"
-	} else {
-		userDef = fmt.Sprintf("tenant/%s", org)
 	}
-	reqUrl := fmt.Sprintf("%s/oauth/%s/token", urlStr, userDef)
-	reqHref, err := url.ParseRequestURI(reqUrl)
+
+	// Create the URL for the register endpoint
+	urlRef, err := url.ParseRequestURI(fmt.Sprintf("%s://%s/oauth/%s/%s", client.VCDHREF.Scheme, client.VCDHREF.Host, userDef, "register"))
 	if err != nil {
-		return nil, fmt.Errorf("error getting request URL from %s : %s", reqUrl, err)
+		return nil, fmt.Errorf("error getting request URL from %s : %s", urlRef.String(), err)
 	}
 
-	data := bytes.NewBufferString(fmt.Sprintf("grant_type=refresh_token&refresh_token=%s", token))
-	req := vcdClient.Client.NewRequest(nil, http.MethodPost, *reqHref, data)
-	req.Header.Add("Accept", "application/*;version=36.1")
+	newTokenParams := &types.ApiTokenParams{}
+	err = client.OpenApiPostItemSync("36.1", urlRef, nil, tokenParams, newTokenParams)
```

Comment:
Could we leverage `getOpenApiHighestElevatedVersion` here based on endpoint to avoid having hardcoded API version

Reply:
It isn't a CloudAPI endpoint, so there's no minimal version. Kept the checks as they are.